### PR TITLE
Preview bubble hover interaction

### PR DESCRIPTION
**Pull request by aosyatnik**
### Purpose

[MAGN-8941](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8941) Make Preview Bubbles more discoverable by removing the preview icon from the node and relying on hover interactions
[MAGN-8943](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8943) User should be able to pin the preview bubble so that it remains opened

Results:
image image

### Declarations

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.

### Reviewers

@ramramps 

### FYIs

@Racel

### CODE REVIEW

File: src/DynamoCoreWpf/Views/Core/NodeView.xaml [hunk 1]
```diff
@@ -281,42 +278,6 @@
                     Height="25"
                     Visibility="{Binding ShouldShowGlyphBar, Converter={StaticResource BooleanToVisibilityCollapsedConverter}}">
 
-            <Canvas x:Name="previewIcon"
```
Comment (aosyatnik) [inline src/DynamoCoreWpf/Views/Core/NodeView.xaml]: This was grey rectangle.

**Comment by Racel** [comment]
@aosyatnik - Can you please remove the search and right arrow button from the toolbar? We will not be working on those for MVP. Please reassign to @ramramps when the PR related to pinning is also ready. I would like them to be merged at the same time.

**Comment by ikeough** [comment]
@aosyatnik You can't imagine how excited this makes me! That little box has been there as long as nodes have been nodes :smiley:

**Comment by aosyatnik** [comment]
@Racel added pin functionality in this PR. Also added pin icons:
image image

**Comment by Racel** [comment]
Thanks, @aosyatnik. Can you try this new icon? I think the unpinned state is hard to read.
Also, can you make sure that the tab and the preview window share the same coincident edge? It looks like that the tab and preview window edges are beside each other instead of right over each other. Hope this makes sense.

unpinned-2px

**Comment by aosyatnik** [comment]
@Racel , I've updated icon:
image

**Comment by Racel** [comment]
@aosyatnik - Thanks@ this looks great!

**Comment by Racel** [comment]
@aosyatnik - Please add the following hover interactions to the pin:
- Change to pointer (hand) cursor when hovering over pin icon
- Change background color of button to #F2F2F2 when hovering over pin icon. See attached.

screen shot 2015-11-12 at 8 28 03 am

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 2]
```diff
@@ -71,6 +74,18 @@ private struct Element
...
         private MirrorData queuedMirrorData;
 
+        /// <summary>
+        ///     Indicates whether preview should stay open, when mouse leaves control.
+        /// </summary>
+        internal bool StaysOpen;
+
+        private const string baseUrl = @"pack://application:,,,/DynamoCoreWpf;component/UI/Images/";
+
+        private readonly BitmapImage pinIconBitmap =
```
Comment (ramramps) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: I recommend using Font Awesome icons here. And use a converter to determine the icon.

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 3]
```diff
@@ -326,7 +335,7 @@ private void RefreshCondensedDisplay(Action refreshDisplay)
                         }
                         else
                         {
-                            newContent = mirrorData.StringData;
+                            newContent = mirrorData.StringData.Substring(0, mirrorData.StringData.IndexOf('('));
```
Comment (ramramps) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: check for empty string here.

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 4]
```diff
@@ -645,14 +661,37 @@ private void OnPreviewControlExpanded(object sender, EventArgs e)
...
+            if (!StaysOpen)
+            {
+                TransitionToState(State.Condensed);
+            }
+        }
+
+        private void OnMapPinMouseClick(object sender, MouseButtonEventArgs e)
+        {
+            StaysOpen = !StaysOpen;
+            PinnIcon.Source = StaysOpen ? pinIconBitmap : unpinIconBitmap;
```
Comment (ramramps) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: Use Converter

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 5]
```diff
@@ -414,16 +424,23 @@ private Size ComputeSmallContentSize()
             this.smallContentGrid.Measure(maxSize);
             Size smallContentGridSize = this.smallContentGrid.DesiredSize;
 
+            // Condensed bubble should be the same width as node or wider.
+            if (smallContentGridSize.Width == 0)
+            {
+                var nodeView = WpfUtilities.FindUpVisualTree<NodeView>(this);
```
Comment (ramramps) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: Check for null.

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs (hunk 2 above)
Comment (Racel) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: @aosyatnik - In order to get rid of unnecessary icons, lets try using FontAwesome icons and varying background colors instead. See below:
pinnedstate-11

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 6]
```diff
@@ -645,14 +676,36 @@ private void OnPreviewControlExpanded(object sender, EventArgs e)
...
             SetCurrentStateAndNotify(State.Condensed);
-            largeContentGrid.Visibility = System.Windows.Visibility.Hidden;
+            largeContentGrid.Visibility = Visibility.Hidden;
+            bubbleTools.Visibility = Visibility.Hidden;
             BeginNextTransition(); // See if there's any more requests.
         }
 
+        private void OnPreviewControlMouseMove(object sender, MouseEventArgs e)
+        {
+            if (IsMouseOver && IsCondensed && !IsInTransition)
```
Comment (ramramps) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: If the state is IsCondensed, then !IsTransition will be true. Do you need 2 checks for this?

**Comment by ramramps** [comment]
Remove the dispatcher timer. @Racel  and I discussed about this. Add a little triangle at the corner of bubblecontrol. When a user hover over the triangle, expand the control. Add the animation in xaml to show the preview bubble for few seconds. I think you can do that using storyboard.
Also,  check the animation. It gets super busy for certain nodes. For ex : add a + node, connect 2 numbers, and check the preview bubble. It shrinks to the width of bubble control. May be you should use DesiredSize instead of actual size. Try that.

**Comment by aosyatnik** [comment]
Hi, @ramramps . I tried to use `StoryBoard`, but that didn't work. If I add `BeginTime`, the animation starts after some time, but these is no opportunity during animation to check whether mouse is over node or not. So, first we have to check mouse position and then fire animation.
However, I found that in .NET 4.5 we can use `async/await`, so I could get rid of dispatcher timer.
Also I fixed issue with short nodes.

Please, take a look, thanks!

**Comment by ramramps** [comment]
Place the mouse pointer just close the node border. Bubble control just hangs continuously

File: src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs [hunk 7]
```diff
@@ -71,6 +74,24 @@ private struct Element
         private bool queuedRefresh;
         private MirrorData queuedMirrorData;
 
+        private static readonly DependencyProperty StaysOpenProperty =
+            DependencyProperty.Register("StaysOpen", typeof(bool), typeof(PreviewControl));
+
+        /// <summary>
+        ///     Indicates whether preview should stay open, when mouse leaves control.
+        /// </summary>
+        public bool StaysOpen
```
Comment (ramramps) [inline src/DynamoCoreWpf/Views/Preview/PreviewControl.xaml.cs]: Make this internal.

**Comment by ramramps** [comment]
Yes. Arrow looks good. Just that hover operation needs some trim there. Otherwise, I am ok there.
But the hovering on nodeview gets tricker when hovered around the node borders. I will show the behavior tomorrow.

**Comment by ramramps** [comment]
@aosyatnik - Remove the mouse leave event on PreviewControl. Fix the .BeginFadeOutTransition on preview control.  @Racel will provide the design for expanding the bubble control.

**Comment by Racel** [comment]
@ramramps - PTAL and let me know if we need to sit together before merging.

**Comment by ramramps** [comment]
@aosyatnik - Check the Number and List.MaximumItem nodes. The values are centered on the mouse hover. Also, for List.MaximumItem, the value got truncated on the hover. 
@Racel said the BubbleControl should be placed in front of all other nodes. Meaning, the z-index should be controlled. otherwise, it is difficult to do the hovering when you have nodes placed very close. Check whether you can do that.

**Comment by Racel** [comment]
@aosyatnik - Also, here is the arrow to replace the one you have in the compact view.
detailarrow

**Comment by aosyatnik** [comment]
@Racel , @ramramps  some results:
image image image image

Didn't get about z-index... It's situation, when one node is above:
image
Mouse is over Number node.

**Comment by Racel** [comment]
@aosyatnik - The new arrow looks great. Let's discuss the z-index issue at our meeting today.
And, a few more styling comments:
- Can you make the height of the pinned icon container the same height as the single line detailed view? See the _SingleFunctionObject image above. They do not match and it is kind of strange.
- Can you please left-align the single line detailed view text?

File: src/DynamoCoreWpf/Views/Preview/WatchTree.xaml [hunk 8]
```diff
@@ -54,10 +54,144 @@
...
+                        </ControlTemplate>
+                    </Setter.Value>
+                </Setter>
+            </Style>
+
             <Style TargetType="{x:Type TreeViewItem}">
-                <Setter Property="IsExpanded" Value="{Binding IsNodeExpanded}">
+                <Setter Property="IsExpanded"
+                        Value="{Binding IsNodeExpanded}" />
+                <Setter Property="Template">
```
Comment (aosyatnik) [inline src/DynamoCoreWpf/Views/Preview/WatchTree.xaml]: I had to rewrite this default TreeViewItem Template in order to get rid of this margin. Actually it's hidden expander, which should be collapsed.
image
image

**Comment by ramramps** [comment]
@aosyatnik  - Run the Graph AttractorWithColor.dyn. then hover over List.MaximumItem node. 
You get a value 2.912... but when you hover over bubble control, the value changes t0 2.913. 
Also the number node. Say you have a number node with value 2. Hover over  the node and then hover over the bubble control. You can see the value "8" changing the position. Please check that.

**Comment by aosyatnik** [comment]
Hi, @ramramps . Number format in expanded view is taken from preference settings:
image

I changed it to "g" format. That means there is no round:
image

Also made Textblocks in WatchTree collapsed, now there is no change in margin in collapsed and expanded state.
image image

However I believe these changes are excessive, because we have separate task for WatchTree improvements. And during this task I had to rewrite not only PreviewControl, but also WatchTree control.

**Comment by ramramps** [comment]
LGTM. Lets get this one merged.

**Comment by Racel** [comment]
Great, @ramramps - please reassign to me for review.

**Comment by aosyatnik** [comment]
New PR: https://github.com/DynamoDS/Dynamo/pull/5779

**Pull request closed without merging**